src/policy/immix/vo_bit_helper.rs

```diff
@@ -0,0 +1,148 @@
+            );
+
+            let mark_bit_meta = mark_bit_spec.extract_side_spec();
+            let vo_bit_meta = vo_bit::VO_BIT_SIDE_METADATA_SPEC;
+
+            assert_eq!(
+                mark_bit_meta.log_bytes_in_region,
+                vo_bit_meta.log_bytes_in_region,
+                "The {s:?} strategy requires the mark bits to have the same granularity as the VO bits."
+            );
+            assert_eq!(mark_bit_meta.log_num_of_bits, vo_bit_meta.log_num_of_bits,
+                "The {s:?} strategy requires the mark bits to have the same number of bits per object as the VO bits.");
+        }
+    }
+}
+
+pub(crate) fn prepare_extra_packets<VM: VMBinding>(
+    chunk_map: &ChunkMap,
+    scheduler: &GCWorkScheduler<VM>,
+) {
+    match strategy::<VM>() {
+        ImmixVOBitUpdateStrategy::ClearAndReconstruct => {
+            // In this strategy, we clear all VO bits after stacks are scanned.
+            let work_packets =
+                chunk_map.generate_tasks(|chunk| Box::new(ClearVOBitsAfterPrepare { chunk }));
+            scheduler.work_buckets[WorkBucketStage::ClearVOBits].bulk_add(work_packets);
+        }
+        ImmixVOBitUpdateStrategy::CopyFromMarkBits => {
+            // Do nothing.
+        }
+    }
+}
+
+pub(crate) fn on_trace_object<VM: VMBinding>(object: ObjectReference) {
+    if strategy::<VM>().vo_bit_available_during_tracing() {
+        // If the VO bits are available during tracing,
+        // we validate the objects we trace using the VO bits.
+        debug_assert!(
+            vo_bit::is_vo_bit_set::<VM>(object),
+            "{:x}: VO bit not set",
+            object
+        );
+    }
+}
+
+pub(crate) fn on_object_marked<VM: VMBinding>(object: ObjectReference) {
+    match strategy::<VM>() {
+        ImmixVOBitUpdateStrategy::ClearAndReconstruct => {
+            // In this strategy, we set the VO bit when an object is marked.
+            vo_bit::set_vo_bit::<VM>(object);
+        }
+        ImmixVOBitUpdateStrategy::CopyFromMarkBits => {
+            // VO bit was not cleared before tracing in this strategy.  Do nothing.
+        }
+    }
+}
+
+pub(crate) fn on_object_forwarded<VM: VMBinding>(new_object: ObjectReference) {
+    match strategy::<VM>() {
+        ImmixVOBitUpdateStrategy::ClearAndReconstruct => {
+            // In this strategy, we set the VO bit of the to-space object when forwarded.
+            vo_bit::set_vo_bit::<VM>(new_object);
+        }
+        ImmixVOBitUpdateStrategy::CopyFromMarkBits => {
+            // In this strategy, we will copy mark bits to VO bits.
+            // We need to set mark bits for to-space objects, too.
+            VM::VMObjectModel::LOCAL_MARK_BIT_SPEC.store_atomic::<VM, u8>(
+                new_object,
+                1,
+                None,
+                Ordering::SeqCst,
+            );
+
+            // In theory, we don't need to set the VO bit for to-space objects because we
```

Comment:
I said that because if a VM only ever visits each edge exactly once, then to-space objects will not need the VO bits set.  For some VMs, such as OpenJDK, this should be easy to guarantee as long as it each object is only enqueued by the GC worker thread that successfully marked or forwarded it.  The situation of Ruby is that currently we rely on the object scanning functions written in C.  We have to call both the "marking function" and the "updating function" for each object in order to both (1) cover all fields and (2) forward as many fields as possible.  That inevitably visits each edge twice.

Perhaps one possible solution is adding a parameter to the `CopyFromMarkBit` enum variant to make it `CopyFromMarkBit(bool)`, where the bool indicates whether to-space objects should have VO bits, too, during tracing.  Currently I don't see such an urgent need for this because Ruby needs it anyway, and I don't see other VMs that don't.  But I'll try to make the comment clearer about why Ruby needs it and what are the alternatives.

Reply:
I was trying to say is that we *have to* set the vo bit for the new object, as the update strategy says vo bits are available during tracing. That means that if we copy an object, the new object is valid, and we should set vo bit for the new object. This has nothing to do with Ruby or any binding, and it has nothing to do with tracing edges once or more than once. It is just about the semantics of the strategy. So I suggest removing the comments that complicate things -- we just set the vo bits here because that what the strategy promised.

Reply:
> ... if we copy an object, the new object is valid...

From the GC's point of view, the new object matches the definition of "an object allocated by MMTk" and is valid.  But from the VM's point of view, suppose we have a VM that visits each edge exactly once.  If it visits an edge and the address contained in the edge happens to be the address of a to-space object, it must be an error, because that means that edge already contained a dangling pointer to an object that is not yet allocated.

But I think it is a very rare corner case.  If a VM needs to use VO bits during tracing, it uses it for some form of conservatism, and conservative fields must pin its children and must be visited before tracing starts (i.e. pinning the children of PPPs).  Allowing to-space objects to have VO bits during tracing will hide some dangling pointers during heap dumping by tracing, but that doesn't matter because heap dumping is a debugging mechanism, and we have ways to know if a reference points to from-space or to-space (though not exposed to the user).

For clearness, I'll just remove the confusing parts of the comment.

Reply:
One of the proposed usage of VO bits is to identify the base reference for internal references. That would happen in tracing, and it is MMTk who uses it rather than the binding. We never claim that VO bits is only for bindings.

Reply:
> One of the proposed usage of VO bits is to identify the base reference for internal references. That would happen in tracing, and it is MMTk who uses it rather than the binding. We never claim that VO bits is only for bindings.

That's right.  In that case, the VM binding should enable some feature such as "interior_pointer" which implicitly requires VO bits to be available during tracing.

A similar feature is heap dumping.  Heap dumping is performed by MMTk core, too.

I think now we have enough reason to just let the VM binding specify if **the binding itself** demands the VO bits to be available during tracing using a boolean flag, and MMTk core can choose the strategy according to that flag as well as enabled features such as heap dumping and interior pointer.

Reply:
> One of the proposed usage of VO bits is to identify the base reference for internal references. That would happen in tracing, and it is MMTk who uses it rather than the binding. We never claim that VO bits is only for bindings.

By the way, when GC starts, interior pointers still point to the interior of from-space objects.  If each edge is visited only once, it will be an error if we find an interior pointer points into the to-space.  It is the same for heap dumping.  If each edge is visited only once, an edge pointing into the to-space should be considered a dangling pointer.

However, we have a flag `PlanConstraints::may_trace_duplicate_edges` and quite some plans have that flag being true.  So I think the VM alone cannot make such a guarantee that each edge is only visited once.  It's safer to just eagerly set the VO bits for to-space objects.

Reply:
I have updated the comment.